### DISCUSSION TITLE
[CHAPTER 3] - 람다 표현식

### DISCUSSION THREAD

**Discussion by corock** [Study-2-Modern-Java-In-Action/Study-2-Modern-Java-in-Action#32] [CHAPTER 3 람다 표현식]
## 3.0 2장에서 확인한 것

- 동작 파라미터화를 이용하면 요구사항에 유연하게 대처 가능
- 정의한 코드 블록을 다른 메서드로 전달할 수 있음
- 정의한 코드 블록을 특정 이벤트가 발생할 때 또는 알고리즘의 일부로 실행되도록 설정할 수 있음

### 문제점

- 코드가 더러움

## 3.1 람다란 무엇인가?

> 메서드로 전달할 수 있는 익명 함수를 단순화한 것

## 3.2 어디에, 어떻게 람다를 사용할까?

> 함수형 인터페이스라는 문맥에서 사용

### 3.2.1 함수형 인터페이스

> 추상 메서드가 1개인 인터페이스

### 3.2.2 함수 디스크럽터(_function descriptor_)

> 함수형 인터페이스의 추상 메서드 시그니처 

## 3.3 람다 활용 : 실행 어라운드 패턴

1. 동작 파라미터화를 기억하라
2. 함수형 인터페이스를 이용해서 동작 전달
3. 동작 실행
4. 람다 전달

## 3.4 함수형 인터페이스 사용

- 자바에서 제공하는 함수형 인터페이스
  - [Package java.util.function](https://docs.oracle.com/javase/8/docs/api/java/util/function/package-summary.html)
  - _cf._ `@FunctionalInterface`

### 3.4.1 `Function<T, R>`

```java

@FunctionalInterface
public interface Function<T, R> {

    /**
     * 이 함수를 주어진 인수에 적용한다.
     *
     * @param t 함수 인수
     * @return 함수 결과
     */
    R apply(T t);

    /**
     * 이 함수를 적용하기 전에 {@code before} 함수를 먼저 적용한 다음 결과에 이 함수를 적용하는 합성 함수를 반환한다.
     * 두 함수 중 하나의 평가가 예외를 발생시키면 합성된 함수의 호출자에게 전달된다.
     *
     * @param <V>    {@code before} 함수와 합성된 함수의 입력 유형 및 합성된 함수의 출력 유형
     * @param before 이 함수가 적용되기 전에 적용할 함수
     * @return 먼저 {@code before} 함수를 적용한 다음 이 함수를 적용하는 합성 함수
     * @throws NullPointerException before 가 null 인 경우
     * @see #andThen(Function)
     */
    default <V> Function<V, R> compose(Function<? super V, ? extends T> before) {
        Objects.requireNonNull(before);
        return (V v) -> apply(before.apply(v));
    }

    /**
     * 이 함수를 먼저 적용한 다음 {@code after} 함수를 결과에 적용하는 합성 함수를 반환한다.
     * 두 함수 중 하나의 평가가 예외를 발생시키면 합성된 함수의 호출자에게 전달된다.
     *
     * @param <V>   {@code after} 함수의 출력 유형 및 합성된 함수의 출력 유형
     * @param after 이 함수가 적용된 후 적용할 함수
     * @return 먼저 이 함수를 적용한 다음 {@code after} 함수를 적용하는 합성 함수
     * @throws NullPointerException after 가 null 인 경우
     * @see #compose(Function)
     */
    default <V> Function<T, V> andThen(Function<? super R, ? extends V> after) {
        Objects.requireNonNull(after);
        return (T t) -> after.apply(apply(t));
    }

    /**
     * 입력 인수를 그대로 반환하는 함수를 반환한다.
     *
     * @param <T> 함수의 입력 및 출력 객체 유형
     * @return 입력 인수를 그대로 반환하는 함수
     */
    static <T> Function<T, T> identity() {
        return t -> t;
    }
}
```

- 나머지 유형도 정의된 문서화 주석을 한번씩 보게 되면 유용
- 기본형에서 특화된 형식의 함수형 인터페이스도 있음
  - `Unary-`: 파라미터 타입과 반환 타입이 같은 경우
  - `Bi-`: 파라미터 인자가 두 개인 경우

### 3.5.1 형식 검사

- _cf._ [Lambda Expression는 어떻게 형식을 검사,추론할까?](https://catsbi.oopy.io/dc24ed56-6ce8-4ed6-b270-cc2b97f4ad5a)

### 형식 검사 과정

1. 람다가 사용된 메서드의 선언을 확인한다.
2. 람다가 사용된 메서드의 파라미터로 대상 형식을 기대한다.
3. 기대하는 파라미터의 함수형 인터페이스를 파악한다.
4. 함수형 인터페이스의 함수 디스크립터를 묘사한다.
5. 전달받은 인수의 람다가 그 요구사항을 만족해야 한다.

### 3.5.2 같은 람다, 다른 함수형 인터페이스

- 예시 1. `Callable<V>` vs. `PrivilegedAction<T>`
- 예시 2. `Comparator<T>` vs. `ToIntBiFunction<T, U>` vs. `BiFunction<T, U, R>`

### 3.5.3 형식 추론

> Q. 형식을 포함하는 것이 좋은가? 형식을 배제하는 것이 좋은가?
>
> A. 알잘딱깔센

#### 예시

```java
Comparator<Apple> byWeight1 = new Comparator<>() {
	public int compare(Apple a1, Apple a2) {
		return a1.getWeight().compareTo(a2.getWeight());
	}
}
Comparator<Apple> byWeight2 = (Apple a1, Apple a2) -> a1.getWeight().compareTo(a2.getWeight());
Comparator<Apple> byWeight3 = (a1, a2) -> a1.getWeight().compareTo(a2.getWeight());
```

### 3.5.4 지역 변수 사용

**자유 변수(_free variable_)**

> 파라미터로 넘겨진 변수가 아닌 외부에서 정의된 변수

**람다 캡처링(_capturing lambda_)**

> 익명 함수가 하는 것처럼 자유 변수를 사용하는 동작

- [Q&A] [람다와 관련된 코드 스니펫 해결하기](https://github.com/orgs/Study-2-Modern-Java-In-Action/discussions/28)

**Comment by JoisFe** [comment]
알잘딱깔센

철지난 유행어 아닌가요?

**Reply by corock** [reply]
철은 다시 돌아올 수도 있습니다. @JoisFe 님에게 노래 하나 추천드려야 겠네요.
[I'd like to return to the classic](https://www.youtube.com/watch?v=DC7ztDaIH8w) 구절에서 깨달음을 얻을 수 있는 곡입니다. :smile:
